ext/drreg/drreg.c

```diff
@@ -1831,115 +1831,194 @@ drreg_is_instr_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT
-                prev_xax_spill = true;
-                if (aflags_in_xax)
-                    aflags_in_xax = false;
-            }
+                aflags_spill_reg = DR_REG_XAX;
+#else
+                aflags_spill_reg = opnd_get_reg(instr_get_dst(inst, 0));
 #endif
+                app_aflags_written_to_reg_now = true;
+            }
+        } else if (instr_get_opcode(inst) ==
+                   IF_X86_ELSE(OP_sahf && last_opcode == OP_cmp, OP_msr)) {
+            if (aflags_spill_reg ==
+                IF_X86_ELSE(DR_REG_XAX, opnd_get_reg(instr_get_src(inst, 0)))) {
+                aflags_native = true;
+                app_aflags_restored_now = true;
+            }
         }
-#ifdef X86
-        else if (prev_xax_spill && instr_get_opcode(&inst) == OP_lahf && spill)
-            aflags_in_xax = true;
-        else if (aflags_in_xax && instr_get_opcode(&inst) == OP_sahf)
-            aflags_in_xax = false;
-#endif
-    }
-    instr_free(drcontext, &inst);
 
-    if (spilled_to_aflags < MAX_SPILLS IF_X86(|| aflags_in_xax)) {
+        /* Mark gpr/aflags as containing native or non-native value based on whether
+         * the current instr is an app or tool write, respectively. Take into account
+         * whether any of these writes were a restore operation, in which case we
+         * skip marking non-native.
+         */
+        for (int i = 0; i < instr_num_dsts(inst); i++) {
+            opnd_t opnd = instr_get_dst(inst, i);
+            if (opnd_is_reg(opnd) && reg_is_gpr(opnd_get_reg(opnd))) {
+                reg_id_t resized_reg = reg_to_pointer_sized(opnd_get_reg(opnd));
+                if (resized_reg == aflags_spill_reg && !app_aflags_written_to_reg_now) {
+                    /* reg does not contain aflags anymore. */
+                    aflags_spill_reg = DR_REG_NULL;
+                }
+                if (instr_is_app(inst)) {
+                    /* App instr wrote reg value. Spill slot value not valid anymore. */
+                    gpr_native[GPR_IDX(resized_reg)] = true;
+                    gpr_spill_slot[GPR_IDX(resized_reg)] = MAX_SPILLS;
+                } else if (!app_gpr_restored_now) {
+                    /* Tool wrote to reg. Not native anymore. */
+                    gpr_native[GPR_IDX(resized_reg)] = false;
+                }
+            }
+        }
+        if (TESTANY(EFLAGS_WRITE_ARITH, instr_get_eflags(inst, DR_QUERY_INCLUDE_ALL))) {
+            if (instr_is_app(inst)) {
+                /* App instr wrote aflags. Saved aflags value (in slot or reg) not valid
+                 * anymore.
+                 */
+                aflags_native = true;
+                aflags_spill_slot = MAX_SPILLS;
+                aflags_spill_reg = DR_REG_NULL;
+            } else if (!app_aflags_restored_now) {
+                /* Tool wrote aflags. Not native anymore. */
+                aflags_native = false;
+            }
+        }
+        last_opcode = instr_get_opcode(inst);
```

Comment:
This should skip labels in case some are added between lahf and seto, breaking the check above.